pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,131 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a function for generating generalized parameter shift rules."""
+import functools
+import warnings
+import numpy as np
+import pennylane as qml
+
+
+@functools.lru_cache(maxsize=None)
+def get_shift_rule(frequencies, shifts=None):
+    r"""Computes the parameter shift rule for a unitary based on its generator's eigenvalue frequency
+     spectrum.
+
+    To compute gradients of circuit parameters in variational quantum algorithms, expressions for
+    cost function first derivatives with respect to the variational parameters can be cast into
+    linear combinations of expectation values at shifted parameter values. These "gradient recipes"
+    can be obtained from the unitary generator's eigenvalue frequency spectrum. Details can be
+    found in https://arxiv.org/abs/2107.12390.
+
+    Args:
+        frequencies (tuple[int or float]): the tuple of eigenvalue frequencies. Eigenvalue
+            frequencies are defined as the unique positive differences obtained from a set of
+            eigenvalues.
+        shifts (tuple[int or float]): the tuple of shift values. If unspecified, equidistant
+            shifts are assumed. Note that if equidistant frequencies are given, any specified
+            shifts will be ignored, and equidistant shifts will be used instead. If supplied,
+            the length of this tuple should match the number of given frequencies.
+
+    Returns:
+         tuple: a tuple of one nested list describing the gradient recipe
+            for the parameter-shift method.
+            This is a tuple with one nested list per operation parameter. For
+            parameter :math:`\phi_k`, the nested list contains elements of the form
+            :math:`[c_i, a_i, s_i]` where :math:`i` is the index of the
+            term, resulting in a gradient recipe of
+
+            .. math:: \frac{\partial}{\partial\phi_k}f = \sum_{i} c_i f(a_i \phi_k + s_i).
+
+
+    Raises:
+        ValueError: if `frequencies` is not a list of unique positive values, or if `shifts`
+            (if specified) is not a list of unique values the same length as `frequencies`.
+
+    **Examples**
+
+    An example of obtaining the frequencies from a set of unique eigenvalues and obtaining the
+    parameter shift rule:
+
+    >>> unique_eigenvals = [1, -1, 0]
+    >>> unique_eigenvals = sorted(unique_eigenvals)
+    >>> frequencies = set()
+    >>> for i in range(len(unique_eigenvals)):
+    >>>     for j in range(i+1, len(unique_eigenvals)):
+    >>>         frequencies.add(unique_eigenvals[j] - unique_eigenvals[i])
+    >>> get_shift_rule(tuple(frequencies))

```

Comment:
I think this can be simplified somewhat:

```suggestion
    >>> unique_eigenvals = [1, 0, -1]
    >>> frequencies = np.diff(sorted(unique_eigenvals))
    >>> frequencies = frequencies[frequencies > 0]
    >>> get_shift_rule(frequencies)
```

Reply:
It looks like `frequencies` comes out as `(1,1)` rather than `(1,2)` since `np.diff` only takes 'nearest neighbour' differences, rather than all pairwise differences. I'm unsure if there's a different function which could be used...

Reply:
It looks like `frequencies` comes out as `(1, 1)` rather than `(1, 2)` since `np.diff` only computes the 'nearest neighbour' differences, rather than all pairwise differences. I'm unsure if theres a different function which could help...

Reply:
Oh yes, good point! The examples I was playing around with locally were too simple, and not catching this edge case 🤦